llm/scripts/validate_paged_llama_model.py

```diff
@@ -0,0 +1,177 @@
+
+from turbine_llm.config import *
+from turbine_llm.data import *
+from turbine_llm.models.llama import *
+
+
+def main(args: list[str]):
+    torch.no_grad().__enter__()
+    config = load_gguf_file(args[0])
+    hp = LlamaHParams.from_gguf_props(config.properties)
+    model = PagedLlamaModelV1(config.root_theta, hp)
+    cache_state = model.cache.allocate(128, torch.float32)
+    start_index = 0
+    next_batch = torch.tensor(
+        [
+            [
+                1,
+                1059,
+                31871,
+                1217,
+                322,
+                266,
+                3682,
+                6075,
+                31902,
+                13,
+                31849,
+                31871,
+                0,
+                0,
+                0,
+                0,
+            ]
+            + 48 * [0],
+            [
+                1,
+                1059,
+                31871,
+                1217,
+                322,
+                31871,
+                0,
+                0,
+                0,
+                0,
+                0,
+                0,
+                0,
+                0,
+                0,
+                0,
+            ]
+            + 48 * [0],
+            64 * [0],
+            64 * [0],
+        ]
+    )
+    assert next_batch.shape[1] % model.cache.block_seq_stride == 0
+    seq_block_ids = torch.tensor(
+        [
+            [127, 0, 0, 0],
+            [126, 0, 0, 0],
+            [0, 0, 0, 0],
+            [0, 0, 0, 0],
+        ]
+    )
+
+    # Important: Do not use a sequence length of 0 for empty batch slots
+    # as it will cause softmax to nan due to a mask of all -inf. This then
+    # propagates and causes badness.
+    seq_lens = torch.tensor([12, 6, 1, 1])
+
+    attention_mask = model.attention_mask(
+        model.input_mask(seq_lens, next_batch.shape[1]),
+        dtype=torch.float32,
+    )
+
+    print(f"Step {start_index}")
+    logits = model.prefill(
+        next_batch,
+        attention_mask=attention_mask,
+        seq_block_ids=seq_block_ids,
+        cache_state=cache_state,
+    )
+    # TODO: Normalize the output of extract_tokens_from_logits into
+    # tensor [bs, 1].
+    tokens = torch.tensor(model.extract_tokens_from_logits(logits, seq_lens)).unsqueeze(
+        1
+    )
+    print(f"  : tokens = {tokens}")
+    print(f"  : cache[127] = {cache_state[0][127]}")
+    print(f"  : cache[126] = {cache_state[0][126]}")
+    print(f"  : cache[0] = {cache_state[0][0]}")
+    print(f"  : cache[1] = {cache_state[0][1]}")
+
+    # Decode a step.
+    print("Decoding...")
+    print(tokens.shape, tokens)
+    start_positions = torch.tensor([12, 6, 0, 0])
+    seq_lens = seq_lens + 1
+    decode_attention_mask = model.decode_attention_mask(
+        model.input_mask(
+            seq_lens,
+            seq_block_ids.shape[1] * model.cache.block_seq_stride,
+        ),
+        dtype=torch.float32,
+    )
+    logits = model.decode(
```

Comment:
QQ: Where do we setup when to release certain sequences. For example what happen if once sequence is finished/found EOS before others?

Reply:
In this text, we don't. I'll end up deleting this test driver once the other one has more mileage on it. This was just the minimal I could do to test prefill and a single decode step.

In the real driver, there has to be an outer scheduler that assigns a sequence to a batch, and it keeps decoding until it observes it can stop (which may mean it overruns eos but the scheduler is responsible for ignoring).

Reply:
sg!